geth/mailservice/mailservice_api.go

```diff
@@ -64,6 +82,90 @@ func setMessagesRequestDefaults(r *MessagesRequest) {
 	}
 }
 
+// waitPeerAdded tries to connect with a peer and waits till connection will be established.
+func (api *PublicAPI) waitPeerAdded(peer *discover.Node, timeout time.Duration) error {
+	log.Debug("waiting to be added", "peer", peer.String())
+	server, err := api.provider.Server()
+	if err != nil {
+		return err
+	}
+	events := make(chan *p2p.PeerEvent, 10)
+	sub := server.SubscribeEvents(events)
+	defer sub.Unsubscribe()
+	server.AddPeer(peer)
+	for {
+		select {
+		case ev := <-events:
+			if ev.Type != p2p.PeerEventTypeAdd {
+				continue
+			}
+			if ev.Peer == peer.ID {
+				return nil
+			}
+		case <-time.After(timeout):
+			return errors.New("failed to add a peer")
+		}
+	}
+}
+
+// trustedPeersGC collects connections that weren't used for defined
+// garbage collector timeout.
+func (api *PublicAPI) trustedPeersGC(timeout, period time.Duration) {
+	connectedPeers := map[*discover.Node]time.Time{}
+	ticker := time.NewTicker(period)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-api.service.quit:
+			return
+		case peer := <-api.newConnectedPeers:
+			connectedPeers[peer] = time.Now()
+		case <-ticker.C:
+			for peer, lastUsed := range connectedPeers {
+				if time.Since(lastUsed) >= timeout {
+					server, err := api.provider.Server()
+					// node was stopped
+					if err != nil {
+						return
```

Comment:
I would print at least a log message there with an error.
Btw, do we have a specific error that the node has stopped? It would be nice to check and to write in a log if anything unexpected is received.
That'd simplify debugging in the future.